### PR TITLE
Fix picking/hover to return closest object

### PR DESCRIPTION
**Pull request by jadh4v**
### EDIT (approach has been changed):
Add `View.pickClosest(xp, yp, maxDistance)` function to select / return a single object that is the closest to the mouse pointer position. Search region is restricted based on the `pointerSize` property. Earlier, the function simply returned/handled the first object in the `selection` array, which was unpredictable. This change affects the 'click' and 'hover' picking modes. The 'select' picking mode (which uses a shift + click-drag a rectangular region) continues to use the previous implementation, which returns all objects within that rectangular region (selection zone) in no particular order.

The algorithm for searching the closest object in the pixel buffer (mentioned [here](https://github.com/Kitware/react-vtk-js/pull/47#issuecomment-975900552)) is already present in vtk-js as [`vtkOpenGLHardwareSelector.getPixelInformation()`](https://github.com/Kitware/vtk-js/blob/c6633f157fd636817f1d638754e58c505fd2b135/Sources/Rendering/OpenGL/HardwareSelector/index.js#L557).

A corresponding fix was required in vtk-js described here: [2263](https://github.com/Kitware/vtk-js/pull/2263).

### CODE REVIEW

File: src/core/View.js [hunk 1]
```diff
@@ -586,29 +586,36 @@ export default class View extends Component {
...
-            compositeID, // Not yet useful unless GlyphRepresentation
-            ...prop.get('representationId'),
-            ray,
-          };
-        })
-        .filter(Boolean);
+      return (
+        this.selections
+          // sort by most dominant objects first.
+          .sort(
```
Comment (jadh4v) [inline src/core/View.js:592]: I simply added the .sort predicate. Other changes are perhaps due to `npm run fix:prettier`.

File: src/core/View.js [hunk 2]
```diff
@@ -586,29 +586,36 @@ export default class View extends Component {
...
+            (a, b) =>
+              b.getProperties().pixelCount - a.getProperties().pixelCount
+          )
+          .map((v) => {
```
Comment (jadh4v) [inline src/core/View.js:596]: I simply added the .sort predicate. Other changes are perhaps due to `npm run fix:prettier`.

**Comment by jadh4v** [comment]
@agirault @floryst
Based on @FezVrasta 's suggestion, we could change the mouse picking to return the closes pixel to current mouse position. The case is trivial where there already is an object at the mouse position. Otherwise, we will return the object that has a foreground pixel closest to current mouse position.

**Comment by jourdain** [comment]
@martinken mentioned to me that on VTK/C++ we have a radius that enable the expected behavior. We should implement it on the JS side.

**Comment by floryst** [comment]
> @martinken mentioned to me that on VTK/C++ we have a radius that enable the expected behavior. We should implement it on the JS side.

Do you have a link to where the C++ implementation is?

**Comment by martinken** [comment]
See https://gitlab.kitware.com/vtk/vtk/-/blob/master/Rendering/Core/vtkHardwareSelector.cxx#L729

Basically it picks from the center, and if nothing is found starts workout out radially until it finds something or hits max distance.

**Comment by jadh4v** [comment]
Cool. This is great. But would this mean that [the selector](https://github.com/Kitware/react-vtk-js/blob/85b897aeaa130a312bc1a673cf0c28bfbc2eeb8d/src/core/View.js#L532) would now only return one result? Or would it be better to implement the new logic as a different function `generateSelection(xp, yp, maxDistance)` ?

**Comment by jadh4v** [comment]
Apologies for this going stale for so long. I will allocate some time soon to resurrect this.

**Comment by jadh4v** [comment]
@floryst @jourdain Ready for review.

File: src/core/View.js [hunk 3]
```diff
@@ -525,6 +521,71 @@ export default class View extends Component {
...
+    const y2 = Math.ceil(yp + tolerance);
+
+    this.selector.setArea(x1, y1, x2, y2);
+    this.previousSelectedData = null;
+
+    if (this.selector.captureBuffers()) {
+      const pos = [xp, yp];
+      const maxDistance = tolerance;
+      const outSelectedPosition = [0, 0];
+      const info = this.selector.getPixelInformation(
```
Comment (jourdain) [inline src/core/View.js:537]: Is that the magic line that will deal with the tolerance?

Comment (jadh4v) [reply src/core/View.js:537]: Yes, the `getPixelInformation` function, which was already present in `vtk-js / vtkOpenGLHardwareSelector`, does a search from pointer position outward within the tolerance zone. Previously `generateSelection` function was used which returns an array of selections within the tolerance footprint. for `pick` and `hover` operations we need the closest actor. Not sure if this generalizes well for other applications!?

Comment (jourdain) [reply src/core/View.js:537]: Maybe we want to pass the expected tolerance. if 0, then it must be the pixel right below the ray.

Comment (floryst) [reply src/core/View.js:537]: Isn't tolerance specified by `maxDistance` here? Or am I missing something?

Comment (floryst) [reply src/core/View.js:537]: Sure thing.

**Comment by jadh4v** [comment]
@jourdain waiting for your green flag. Also I don't have write privileges, so someone will have to merge for me.

**Pull request merged**